sway-core/src/language/ty/declaration/constant.rs

```diff
@@ -16,37 +17,47 @@ pub struct TyConstantDeclaration {
     pub visibility: Visibility,
     pub is_configurable: bool,
     pub attributes: transform::AttributesMap,
+    pub return_type: TypeId,
     pub type_ascription: TypeArgument,
     pub span: Span,
+    pub implementing_type: Option<TyDeclaration>,
 }
 
 impl EqWithEngines for TyConstantDeclaration {}
 impl PartialEqWithEngines for TyConstantDeclaration {
     fn eq(&self, other: &Self, engines: Engines<'_>) -> bool {
+        let type_engine = engines.te();
         self.call_path == other.call_path
             && self.value.eq(&other.value, engines)
             && self.visibility == other.visibility
             && self.type_ascription.eq(&other.type_ascription, engines)
             && self.is_configurable == other.is_configurable
+            && type_engine
+                .get(self.return_type)
+                .eq(&type_engine.get(other.return_type), engines)
     }
 }
 
 impl HashWithEngines for TyConstantDeclaration {
     fn hash<H: Hasher>(&self, state: &mut H, engines: Engines<'_>) {
+        let type_engine = engines.te();
         let TyConstantDeclaration {
             call_path,
             value,
             visibility,
+            return_type,
             type_ascription,
             is_configurable,
             // these fields are not hashed because they aren't relevant/a
             // reliable source of obj v. obj distinction
             attributes: _,
             span: _,
+            implementing_type: _,
         } = self;
         call_path.hash(state);
         value.hash(state, engines);
         visibility.hash(state);
+        type_engine.get(*return_type).hash(state, engines);
```

Comment:
Same question here.. What about `implementing_type`? Can't we have two identical constants but implemented differently for two different types?